staging/src/k8s.io/client-go/tools/cache/wait_test.go

```diff
@@ -0,0 +1,246 @@
+			checkers:   []DoneChecker{newMockChecker("first", 10*time.Second), newMockChecker("second", -1), newMockChecker("last", 0*time.Second)},
+			expectDone: false,
+			expectOutput: `I0101 00:00:00.000000 %7[1]d wait_test.go:%[2]d] "Waiting" for="my-caches"
+I0101 00:00:00.000000 %7[1]d wait_test.go:%[2]d] "Done waiting" for="my-caches" instance="last"
+I0101 00:00:10.000000 %7[1]d wait_test.go:%[2]d] "Done waiting" for="my-caches" instance="first"
+I0101 00:01:00.000000 %7[1]d wait_test.go:%[2]d] "Timed out waiting" for="my-caches" cause="context deadline exceeded" instances=["second"]
+`,
+		},
+		"some-canceled": {
+			timeout:    -1,
+			what:       "my-caches",
+			checkers:   []DoneChecker{newMockChecker("first", 10*time.Second), newMockChecker("second", -1), newMockChecker("last", 0*time.Second)},
+			expectDone: false,
+			expectOutput: `I0101 00:00:00.000000 %7[1]d wait_test.go:%[2]d] "Waiting" for="my-caches"
+I0101 00:00:00.000000 %7[1]d wait_test.go:%[2]d] "Done waiting" for="my-caches" instance="last"
+I0101 00:00:00.000000 %7[1]d wait_test.go:%[2]d] "Timed out waiting" for="my-caches" cause="context canceled" instances=["first","second"]
+`,
+		},
+		"more": {
+			timeoutReason: "go fish",
+			timeout:       5 * time.Second,
+			what:          "my-caches",
+			checkers:      []DoneChecker{newMockChecker("first", 10*time.Second), newMockChecker("second", -1), newMockChecker("last", 0*time.Second)},
+			expectDone:    false,
+			expectOutput: `I0101 00:00:00.000000 %7[1]d wait_test.go:%[2]d] "Waiting" for="my-caches"
+I0101 00:00:00.000000 %7[1]d wait_test.go:%[2]d] "Done waiting" for="my-caches" instance="last"
+I0101 00:00:05.000000 %7[1]d wait_test.go:%[2]d] "Timed out waiting" for="my-caches" cause="go fish" instances=["first","second"]
+`,
+		},
+		"all": {
+			timeout:    time.Minute,
+			what:       "my-caches",
+			checkers:   []DoneChecker{newMockChecker("first", -1), newMockChecker("second", -1), newMockChecker("last", -1)},
+			expectDone: false,
+			expectOutput: `I0101 00:00:00.000000 %7[1]d wait_test.go:%[2]d] "Waiting" for="my-caches"
+I0101 00:01:00.000000 %7[1]d wait_test.go:%[2]d] "Timed out waiting" for="my-caches" cause="context deadline exceeded" instances=["first","last","second"]
+`,
+		},
+	} {
+		t.Run(name, func(t *testing.T) {
+			synctest.Test(t, func(t *testing.T) {
+				var buffer bytes.Buffer
+				logger := textlogger.NewLogger(textlogger.NewConfig(textlogger.Output(&buffer)))
+				ctx := klog.NewContext(context.Background(), logger)
+				var wg sync.WaitGroup
+				defer wg.Wait()
+				if tc.timeout != 0 {
+					switch tc.timeoutReason {
+					case "":
+						if tc.timeout > 0 {
+							c, cancel := context.WithTimeout(ctx, tc.timeout)
+							defer cancel()
+							ctx = c
+						} else {
+							c, cancel := context.WithCancel(ctx)
+							cancel()
+							ctx = c
+						}
+					default:
+						c, cancel := context.WithCancelCause(ctx)
+						wg.Go(func() {
+							time.Sleep(tc.timeout)
```

Comment:
Is there no way to test this without sleeping on the timeout? Just wondering for time to complete on test.

Reply:
Gotcha, just have it engrained to flag sleeps whenever I see them. Thanks!